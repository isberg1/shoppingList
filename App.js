import React from 'react';
import {SafeAreaView, View, StatusBar} from 'react-native';
import {ShoppingList} from './ShoppingList/ShoppingList';
import {styles} from './appStyles';

const App = () => {
  return (
    <View style={styles.root}>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView style={styles.safeAreaView} />
      <ShoppingList />
    </View>
  );
};

export default App;
