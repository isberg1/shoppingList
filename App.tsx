/* *
 * @format
 */

import React from 'react';
import {SafeAreaView, View, StatusBar} from 'react-native';
import {ShoppingList} from './ShoppingList/ShoppingList';
import context from './ShoppingList/Context';
import useSettingsManager from './ShoppingList/Hooks/UseSettingsManager';
import {styles} from './appStyles';
import {SteamBlue} from './ShoppingList/Components/commonStyles';

const App = () => {
  const settings = useSettingsManager();

  return (
    <context.Provider value={{...settings}}>
      <View style={styles.root}>
        <StatusBar backgroundColor={SteamBlue} />
        <SafeAreaView style={styles.safeAreaView} />
        <ShoppingList />
      </View>
    </context.Provider>
  );
};

export default App;
