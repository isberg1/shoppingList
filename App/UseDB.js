import React, {useState, useEffect, useCallback, useMemo} from 'react';
import AsyncStorage from '@react-native-community/async-storage';

const LIST_KEY = 'LIST';

export const useDB = () => {
  const [list, setList] = useState([]);

  useEffect(() => {
    const _retrieveData = async () => {
      try {
        const value = await AsyncStorage.getItem(LIST_KEY);
        if (value !== null) {
          // We have data!!
          const val = JSON.parse(value);
          console.log('retrieveData', val);
          setList(val);
        }
      } catch (error) {
        // Error retrieving data
      }
    };
    _retrieveData();
  }, []);

  useEffect(() => {
    if (list && list.length > 0) {
      const _storeData = async () => {
        try {
          await AsyncStorage.setItem(LIST_KEY, JSON.stringify(list));
          console.log('storeData');
        } catch (error) {
          console.log('store data error', error);
        }
      };
      _storeData();
    }
  }, [list]);

  const _deleteData = async () => {
    try {
      await AsyncStorage.removeItem(LIST_KEY);
      console.log('deleteData');
    } catch (error) {
      console.log('delete data error', error);
    }
  };

  const addToList = item => {
    setList(val => [...val, item]);
  };

  const deleteList = item => {
    setList([]);
    _deleteData();
  };

  return [list, addToList, deleteList];
};
