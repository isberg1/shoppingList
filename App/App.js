import React, {useState, useCallback, useContext, useMemo} from 'react';
import {SafeAreaView, View, StatusBar} from 'react-native';

import {styles} from './styles';
import {Buttons, List, Input} from '../Components/index';
import {useDB} from './UseDB';

export const MyContext = React.createContext({
  num: 0,
});

export const modes = {
  add: 'add',
  edit: 'edit',
  delete: 'delete',
};

let storeLastEditedIndex;

const App = () => {
  const [inputValue, setInputValue] = useState('');
  const {list, addToList, deleteList, removeItem, editList} = useDB({});
  const [itemIsTouched, setItemIsTouched] = useState({});
  const [mode, setMode] = useState(modes.add);
  const contextObject = useContext(MyContext);

  const itemIsTouchedLength = useCallback(
    () => Object.keys(itemIsTouched).length,
    [itemIsTouched],
  );
  const onPressAdd = () => {
    if (inputValue) {
      addToList(inputValue);
      setInputValue('');
      if (itemIsTouchedLength()) setMode(modes.delete);
    }
  };

  const onPressEdit = () => {
    if (inputValue && mode === modes.edit) {
      editList(storeLastEditedIndex, inputValue);
      setInputValue('');
      setMode(modes.add);
    }
  };

  const onPressDelete = () => {
    if (!Object.keys(itemIsTouched).length) return;

    const toDel = Object.keys(itemIsTouched)
      .filter(val => itemIsTouched[val] === true)
      .map(val => parseInt(val, 10));

    removeItem(toDel);
    setItemIsTouched({});
    setMode(modes.add);
    contextObject.num = contextObject.num + 'g';
  };

  const _onPressList = (index, value) => {
    const newItemIsTouched = itemIsTouched;
    newItemIsTouched[index] = value;
    if (value) {
      setItemIsTouched(newItemIsTouched);
    } else {
      delete newItemIsTouched[index];
      setItemIsTouched(newItemIsTouched);
    }

    if (itemIsTouchedLength() > 0 && value && !inputValue) {
      setMode(modes.delete);
    } else if (itemIsTouchedLength() === 0 && !value) {
      setMode(modes.add);
    }
  };

  const _onLongPressList = index => {
    setInputValue(list[index]);
    setMode(modes.edit);
    storeLastEditedIndex = index;
  };

  const buttonDisabled = useMemo(() => {
    switch (mode) {
      case modes.add:
        return !!inputValue;
      case modes.delete:
        return !!itemIsTouchedLength();
      case modes.edit:
        return !!inputValue;
    }
  }, [inputValue, itemIsTouchedLength, mode]);

  const inputHandler = text => {
    if (mode === modes.delete) setMode(modes.add);
    setInputValue(text);
  };

  return (
    <MyContext.Provider value={contextObject}>
      <View style={styles.root}>
        <StatusBar barStyle="dark-content" />
        <SafeAreaView style={styles.safeAreaView}>
          <Input value={inputValue} onChangeText={inputHandler} />
          <Buttons
            disabled={!buttonDisabled}
            onPressAdd={onPressAdd}
            onPressDelete={onPressDelete}
            onPressEdit={onPressEdit}
            mode={mode}
          />
          <List
            list={list}
            onPress={_onPressList}
            onLongPress={_onLongPressList}
          />
        </SafeAreaView>
      </View>
    </MyContext.Provider>
  );
};

export default App;
