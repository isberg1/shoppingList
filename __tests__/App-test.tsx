/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../App';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('@react-native-community/async-storage', () => ({
  setItem: jest.fn(),
  removeItem: jest.fn(),
}));

it('renders correctly', () => {
  renderer.create(<App />);
});
