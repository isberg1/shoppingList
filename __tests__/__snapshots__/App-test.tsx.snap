// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches Snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#1b2838",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "height": "100%",
        "marginTop": 40,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "marginBottom": 20,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#5A5A5A",
              "borderRadius": 5,
              "borderWidth": 2,
              "flexDirection": "row",
              "marginHorizontal": 30,
              "marginVertical": 10,
              "opacity": 0.7,
            },
            false,
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          clearButtonMode="never"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Enter Text"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "white",
              "flex": 8,
              "fontSize": 30,
              "padding": 5,
              "textAlign": "center",
              "textDecorationLine": "underline",
            }
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
                "paddingHorizontal": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 25,
                }
              }
            >
              X
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 5,
              "borderWidth": 2,
              "fontSize": 50,
              "marginHorizontal": 30,
            },
            Object {
              "opacity": 0.6,
            },
            Object {
              "backgroundColor": "green",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0)",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 20,
                    "height": 60,
                    "textAlign": "center",
                    "textAlignVertical": "center",
                  },
                ]
              }
            >
              Add Item
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 4,
          "paddingBottom": 80,
          "paddingTop": 20,
        }
      }
    >
      <RCTScrollView
        contentInsetAdjustmentBehavior="automatic"
        style={
          Object {
            "marginHorizontal": 30,
          }
        }
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</View>
`;
